<a href="https://colab.research.google.com/github/BEST-PK/DPDM2024/blob/main/DPDM2024_CH6_Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import os

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
path = '/content/drive/MyDrive/dpdm24_DATA/reduced_marketbasket.csv'

In [ ]:
#data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [6]:
data = pd.read_csv(path,encoding='latin-1')

In [8]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [15]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [24]:
transactions= data.groupby('InvoiceNo')['StockCode'].apply(list)  #.reset_index()
transactions

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

## Association Rules

In [3]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5953 sha256=0e001f07b8d80561f6d182646faedd7012ac5e47514e06344d4339e8a52606d1
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [4]:
from apyori import apriori

In [ ]:
list(apriori(transactions,min_support = 0.03))

In [38]:
frequentItemsets = list(apriori(transactions,min_support = 0.03))
frequentItemsets

[RelationRecord(items=frozenset({'20712'}), support=0.033667953667953665, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.033667953667953665, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.03254826254826255, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.03254826254826255, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04054054054054054, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04054054054054054, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.062084942084942085, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.062084942084942085, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.04003861003861004, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozense

In [39]:
len(frequentItemsets)

80

In [40]:
frequentItemsets[0][0]

frozenset({'20712'})

In [51]:
frequentItemsets[-1][0]

frozenset({'22697', '22699'})

In [43]:
frequentItemsets[0] #RelationRecord 1

RelationRecord(items=frozenset({'20712'}), support=0.033667953667953665, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.033667953667953665, lift=1.0)])

In [44]:
frequentItemsets[-1] #RelationRecord 80 (สุดท้าย)

RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [50]:
dir(frequentItemsets[-1][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [52]:
frequentItemsets[-1][0].__len__()

2

**ใช้ฟังก์ชันเพื่อหามากกว่า 1 items**

In [54]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


#**Result 2 items**

RelationRecord(items=frozenset({'85099B', '22386'}),

support=0.03216216216216216,

Ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0),



*   OrderedStatistic(items_base=frozenset({'22386'}),items_add=frozenset({'85099B'}),` confidence=0.6766856214459789` , lift=8.20897311262335),
*   OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])






---

RelationRecord(items=frozenset({'22697', '22699'}),

 support=0.03027027027027027,

  Ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0),

  
*    OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), `confidence=0.7417218543046358`, lift=17.1523178807947),
*   OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])
  
   
   
   

In [ ]:
set(data[(data['StockCode']=='22386')]['Description'])

In [64]:
set(data[(data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG RED RETROSPOT'}

In [65]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

In [66]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER', 'ROSES REGENCY TEACUP AND SAUCER '}

In [67]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'C558312',
 '550638',
 '578993',
 '557605',
 '549016',
 '538153',
 '542999',
 '560473',
 '575866',
 '558923',
 '551599',
 '544186',
 '565214',
 'C538124',
 '551058',
 '552663',
 '579518',
 '552336',
 'C572410',
 'C543837',
 '553612',
 '567368',
 '545377',
 '555486',
 '573125',
 '539752',
 '544887',
 '547206',
 '541130',
 '567793',
 '570701',
 'C537853',
 'C566758',
 '566581',
 '572552',
 '556205',
 '559052',
 '536524',
 '575063',
 '543036',
 '546545',
 '547547',
 '566988',
 '566720',
 '551387',
 '556073',
 '550499',
 '556577',
 '568060',
 '539737',
 '561698',
 '551171',
 '577298',
 '540352',
 '565155',
 '552259',
 '546892',
 '546683',
 '557034',
 '566124',
 '577039',
 '559055',
 '557859',
 '571931',
 '554329',
 '554921',
 '548522',
 '541811',
 'C541822',
 '561702',
 '537176',
 '561894',
 '547936',
 '544681',
 'C553479',
 'C564212',
 '571241',
 '550978',
 '566326',
 '556783',
 '543251',
 '543546',
 '566477',
 '565779',
 '557645',
 '565223',
 '555722',
 '539465',
 '566236',
 '579751',
 

In [ ]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

# **HW 2 หา pattern ที่น่าสนใจในหนังอินเดีย**

สนใจ Director, Genre ในการนำมาหา pattern ที่น่าสนใจในหนังอินเดีย
"ผู้กำกับสร้างหนังแนวไหนออกมาฉายคู่กันบ้าง ?"

In [31]:
movies_india = pd.read_csv('/content/drive/MyDrive/dpdm24_DATA/IMDb Movies India.csv' , encoding='latin-1')
movies_india.head()

Name    Year Duration            Genre  \
0                                         NaN      NaN            Drama   
1  #Gadhvi (He thought he was Gandhi)  (2019)  109 min            Drama   
2                         #Homecoming  (2021)   90 min   Drama, Musical   
3                             #Yaaram  (2019)  110 min  Comedy, Romance   
4                   ...And Once Again  (2010)  105 min            Drama   

   Rating Votes            Director       Actor 1             Actor 2  \
0     NaN   NaN       J.S. Randhawa      Manmauji              Birbal   
1     7.0     8       Gaurav Bakshi  Rasika Dugal      Vivek Ghamande   
2     NaN   NaN  Soumyajit Majumdar  Sayani Gupta   Plabita Borthakur   
3     4.4    35          Ovais Khan       Prateik          Ishita Raj   
4     NaN   NaN        Amol Palekar  Rajat Kapoor  Rituparna Sengupta   

           Actor 3  
0  Rajendra Bhatia  
1    Arvind Jangid  
2       Roy Angana  
3  Siddhant Kapoor  
4      Antara Mali

In [32]:
movies_india = movies_india[['Director','Genre']]
movies_india

Director            Genre
0           J.S. Randhawa            Drama
1           Gaurav Bakshi            Drama
2      Soumyajit Majumdar   Drama, Musical
3              Ovais Khan  Comedy, Romance
4            Amol Palekar            Drama
...                   ...              ...
15504       Mahendra Shah           Action
15505          Kuku Kohli    Action, Drama
15506          Kiran Thej           Action
15507                 NaN           Action
15508        K.C. Bokadia    Action, Drama

[15509 rows x 2 columns]

**การตรวจสอบค่าว่าง isnull()**

In [33]:
movies_india[['Director', 'Genre']].isnull().any() # Check for nulls in the specified columns.

Director    True
Genre       True
dtype: bool

จัดการค่าว่าง

โดยลบออก เฉพาะ column ที่เราสนใจคือ Director, Genre ในการนำมาหา pattern

In [34]:
movies_india = movies_india.dropna(subset=['Director','Genre'])
movies_india

Director                 Genre
0           J.S. Randhawa                 Drama
1           Gaurav Bakshi                 Drama
2      Soumyajit Majumdar        Drama, Musical
3              Ovais Khan       Comedy, Romance
4            Amol Palekar                 Drama
...                   ...                   ...
15503     S.P. Muthuraman  Action, Crime, Drama
15504       Mahendra Shah                Action
15505          Kuku Kohli         Action, Drama
15506          Kiran Thej                Action
15508        K.C. Bokadia         Action, Drama

[13131 rows x 2 columns]

In [35]:
movies_india.groupby('Director')['Genre'].apply(set).apply(list)

Director
A B Arjun                                                [Thriller]
A. Bhimsingh      [Drama, Action, Action, Comedy, Drama, Biograp...
A. Habib                                                    [Drama]
A. Jagannathan    [Action, Drama, Action, Drama, Romance, Crime,...
A. Karim                                            [Drama, Family]
                                        ...                        
Ziaullah Khan                                       [Drama, Comedy]
Zoya Akhtar       [Drama, Drama, Music, Romance, Drama, Romance,...
Zubair Khan                                         [Crime, Horror]
Zul Vellani                                                 [Drama]
Zunaid Memon                                [Crime, Drama, Mystery]
Name: Genre, Length: 5599, dtype: object

In [36]:
#transactions1= movies_india.groupby('Year')['Genre'].apply(set).apply(list)  #.reset_index()
#transactions1
transactions_india_movies= movies_india.groupby('Director')['Genre'].apply(set).apply(list)  #.reset_index()
transactions_india_movies

Director
A B Arjun                                                [Thriller]
A. Bhimsingh      [Drama, Action, Action, Comedy, Drama, Biograp...
A. Habib                                                    [Drama]
A. Jagannathan    [Action, Drama, Action, Drama, Romance, Crime,...
A. Karim                                            [Drama, Family]
                                        ...                        
Ziaullah Khan                                       [Drama, Comedy]
Zoya Akhtar       [Drama, Drama, Music, Romance, Drama, Romance,...
Zubair Khan                                         [Crime, Horror]
Zul Vellani                                                 [Drama]
Zunaid Memon                                [Crime, Drama, Mystery]
Name: Genre, Length: 5599, dtype: object

**วิเคราะห์ Parterns**

In [37]:
!pip install apyori
import pandas as pd
from apyori import apriori

In [58]:
frequentItemsets = list(apriori(transactions_india_movies,min_support = 0.03))
frequentItemsets

[RelationRecord(items=frozenset({'Action'}), support=0.14038221110912663, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Action'}), confidence=0.14038221110912663, lift=1.0)]),
 RelationRecord(items=frozenset({'Action, Crime, Drama'}), support=0.05679585640292909, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Action, Crime, Drama'}), confidence=0.05679585640292909, lift=1.0)]),
 RelationRecord(items=frozenset({'Action, Drama'}), support=0.04072155742096803, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Action, Drama'}), confidence=0.04072155742096803, lift=1.0)]),
 RelationRecord(items=frozenset({'Comedy'}), support=0.0773352384354349, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Comedy'}), confidence=0.0773352384354349, lift=1.0)]),
 RelationRecord(items=frozenset({'Comedy, Drama'}), support=0.04036435077692445, ordered_statistics=[Ordered

In [59]:
len(frequentItemsets)

16

In [40]:
min_sup = 0.03

frequentItemsets_2 = list(apriori(transactions_india_movies,min_support = min_sup))
for i in frequentItemsets_2:
    if len(i[0])>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'Action', 'Drama'}), support=0.03714949098053224, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Action', 'Drama'}), confidence=0.03714949098053224, lift=1.0), OrderedStatistic(items_base=frozenset({'Action'}), items_add=frozenset({'Drama'}), confidence=0.26463104325699743, lift=0.7906452567747752), OrderedStatistic(items_base=frozenset({'Drama'}), items_add=frozenset({'Action'}), confidence=0.11099252934898612, lift=0.7906452567747752)])
yes
RelationRecord(items=frozenset({'Drama, Romance', 'Drama'}), support=0.03196999464190034, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Drama, Romance', 'Drama'}), confidence=0.03196999464190034, lift=1.0), OrderedStatistic(items_base=frozenset({'Drama'}), items_add=frozenset({'Drama, Romance'}), confidence=0.09551760939167556, lift=1.317249002423624), OrderedStatistic(items_base=frozenset({'Drama, Romance'}), items_add=frozenset({'Drama'})

**Result 2 items**


RelationRecord(items=frozenset`({'Action', 'Drama'}), ` : ผู้กำกับจะสร้างหนังประเภท 'Action' และ 'Drama' ออกมาฉายคู่กัน

support=0.03714949098053224,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Action', 'Drama'}), confidence=0.03714949098053224, lift=1.0),

    *OrderedStatistic(items_base=frozenset({'Action'}), items_add=frozenset({'Drama'}), confidence=0.26463104325699743, lift=0.7906452567747752),
    **OrderedStatistic(items_base=frozenset({'Drama'}), items_add=frozenset({'Action'}), confidence=0.11099252934898612, lift=0.7906452567747752)])




---


RelationRecord(items=frozenset`({'Drama, Romance', 'Drama'}),` : ผู้กำกับจะสร้างหนังประเภท 'Drama, Romance' และ 'Drama' ออกมาฉายคู่กัน

support=0.03196999464190034,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Drama, Romance', 'Drama'}), confidence=0.03196999464190034, lift=1.0),

    *OrderedStatistic(items_base=frozenset({'Drama'}), items_add=frozenset({'Drama, Romance'}), confidence=0.09551760939167556, lift=1.317249002423624),
    *OrderedStatistic(items_base=frozenset({'Drama, Romance'}), items_add=frozenset({'Drama'}), confidence=0.4408866995073891, lift=1.317249002423624)])



---


RelationRecord(items=frozenset`({'Romance', 'Drama'}),`:ผู้กำกับจะสร้างหนังประเภท 'Romance' และ 'Drama' ออกมาฉายคู่กัน

 support=0.030719771387747813,

 ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Romance', 'Drama'}), confidence=0.030719771387747813, lift=1.0),

    *OrderedStatistic(items_base=frozenset({'Drama'}), items_add=frozenset({'Romance'}), confidence=0.09178228388473852, lift=0.9429156100378918),
    *OrderedStatistic(items_base=frozenset({'Romance'}), items_add=frozenset({'Drama'}), confidence=0.3155963302752294, lift=0.9429156100378918)])

ลองทำ ปี กับ ประเภทหนัง

In [48]:
movies_india_1 = pd.read_csv('/content/drive/MyDrive/dpdm24_DATA/IMDb Movies India.csv' , encoding='latin-1')


In [53]:
movies_india_1 = movies_india_1[['Year','Genre']]
movies_india_1 = movies_india_1.dropna(subset=['Year','Genre'])

In [54]:
transactions_india_movies_1= movies_india_1.groupby('Year')['Genre'].apply(set).apply(list)  #.reset_index()
transactions_india_movies_1

Year
(1917)                          [Adventure, Drama, Fantasy]
(1922)                             [Action, Drama, History]
(1928)                             [Action, Drama, History]
(1931)    [Action, Adventure, Fantasy, Musical, Drama, R...
(1932)        [Adventure, Fantasy, Musical, Romance, Drama]
                                ...                        
(2018)    [History, Drama, Sport, Sport, Biography, Musi...
(2019)    [Animation, Action, Adventure, Action, Adventu...
(2020)    [History, Drama, Sport, War, Drama, Musical, D...
(2021)    [Musical, Romance, Drama, Sport, Comedy, Music...
(2022)    [Biography, Horror, Musical, Thriller, Drama, ...
Name: Genre, Length: 95, dtype: object

In [65]:
S=list(apriori(transactions_india_movies_1,min_support = 0.7))
S

[RelationRecord(items=frozenset({'Action'}), support=0.8842105263157894, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Action'}), confidence=0.8842105263157894, lift=1.0)]),
 RelationRecord(items=frozenset({'Comedy'}), support=0.8315789473684211, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Comedy'}), confidence=0.8315789473684211, lift=1.0)]),
 RelationRecord(items=frozenset({'Drama'}), support=0.968421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Drama'}), confidence=0.968421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'Drama, Family'}), support=0.8421052631578947, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Drama, Family'}), confidence=0.8421052631578947, lift=1.0)]),
 RelationRecord(items=frozenset({'Drama, Romance'}), support=0.8631578947368421, ordered_statistics=[OrderedStatistic(items_base=frozenset(), item

In [64]:
len(S)

267